staging/src/k8s.io/client-go/tools/cache/shared_informer.go

```diff
@@ -280,45 +345,106 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
 	for _, d := range obj.(Deltas) {
 		switch d.Type {
 		case Sync, Added, Updated:
+			isSync := d.Type == Sync
 			s.cacheMutationDetector.AddObject(d.Object)
 			if old, exists, err := s.indexer.Get(d.Object); err == nil && exists {
 				if err := s.indexer.Update(d.Object); err != nil {
 					return err
 				}
-				s.processor.distribute(updateNotification{oldObj: old, newObj: d.Object})
+				s.processor.distribute(updateNotification{oldObj: old, newObj: d.Object}, isSync)
 			} else {
 				if err := s.indexer.Add(d.Object); err != nil {
 					return err
 				}
-				s.processor.distribute(addNotification{newObj: d.Object})
+				s.processor.distribute(addNotification{newObj: d.Object}, isSync)
 			}
 		case Deleted:
 			if err := s.indexer.Delete(d.Object); err != nil {
 				return err
 			}
-			s.processor.distribute(deleteNotification{oldObj: d.Object})
+			s.processor.distribute(deleteNotification{oldObj: d.Object}, false)
 		}
 	}
 	return nil
 }
 
 type sharedProcessor struct {
-	listeners []*processorListener
+	listenersLock sync.RWMutex
+	listeners     []*processorListener
+	clock         clock.Clock
+}
+
+func (p *sharedProcessor) addListener(listener *processorListener) {
+	p.listenersLock.Lock()
+	defer p.listenersLock.Unlock()
+
+	p.listeners = append(p.listeners, listener)
 }
 
-func (p *sharedProcessor) distribute(obj interface{}) {
+func (p *sharedProcessor) distribute(obj interface{}, isSync bool) {
+	p.listenersLock.RLock()
+	defer p.listenersLock.RUnlock()
+
 	for _, listener := range p.listeners {
+		if isSync && !listener.shouldHandleSyncItem() {
```

Comment:
You're saying that the shared informer could be a slow consumer of the queue, the queue has items from resync *n*, and while the informer is slowly popping, resync *n+1* happens?